### DISCUSSION TITLE
HM Programme aktivieren/deaktivieren

### DISCUSSION THREAD

**Discussion by pdmetze** [SukramJ/aiohomematic#1938] [Ideas]
Hi, es wäre klasse, wenn man über Home Assistant die Homematic Programme aktivieren oder deaktivieren kann. Damit meine ich nicht, ein Programm auszuführen, was ja bereits geht. In der Homematic gibt es die Möglichkeit unter Programme, sie ganz rechts mit dem Häkchen "active" zu aktivieren oder zu deaktivieren.
![homematic_programs](https://github.com/user-attachments/assets/6dd05a70-5fb6-4a34-a0be-cae8ecb8fbf4)

Hintergrund:
Ich nutze diese Programme für Homematic interne/systemnahe Logiken, beispielsweise steuere ich dadurch das Licht über Bewegungsmelder. Nun möchte ich Bewegungsmelder aber auch mal "deaktivieren", und das mache ich ganz einfach indem ich das zugehörige Programm deaktiviere. Ich komme ganz frisch von ioBroker, dort gibt es einfach ein boolean Objekt namens "programm.Active", das ich auf true/false setzen kann. Damit aktiviere ich dann in Homematic das Häkchen.

Zusammengefasst:
Es gibt Logiken die man in der Homematic abbilden will, um sie stabiler gegen Ausfälle zu haben. HM Programme sind ja nicht nur Befehlsabfolgen, die man selbst aktiv ausführt, sondern haben eigene Trigger und Bedingungen und werden anhand derer ausgeführt. Über Home Assistant Dashboards möchte man nun aber möglicherweise die Ausführung für eine bestimmte Zeit deaktivieren oder aktivieren.

/edit:
ich bin ganz neu in Home Assistant und weiß nicht, ob ich die datenstrukturen richtig verstehe. aber könnte man das vielleicht so lösen, dass man im Home Assistant zur entity des programms ein attribut für aktiv/inaktiv hinzufügt, und es über services aktivieren oder deaktivieren kann?

**Comment by SukramJ** [comment]
Ich kenne keine bestehende API, über die man Programme aktivieren oder deaktivieren kann.

**Reply by pdmetze** [reply]
Bin selbst leider nicht versiert genug um dir genaueres zu sagen, aber ich habe dazu gefunden, dass das wohl über die ReGaHSS-API bzw. Homematic Skript API gehen soll.

Laut ChatGPT:
> In Homematic kannst du Programme über die ReGaHSS-API (Homematic Script-API) aktivieren oder deaktivieren. Die ReGaHSS-API ermöglicht es, auf Programme, Variablen und andere Elemente der Homematic-Zentrale zuzugreifen und sie zu steuern.
> 
> Hier ist ein Überblick, wie du Programme aktivieren oder deaktivieren kannst:
> 
> 1. Zugriff auf die ReGaHSS-API
> 
> Die API ist über den HTTP-Endpunkt der Homematic CCU oder kompatibler Systeme (z. B. RaspberryMatic) erreichbar. Typischerweise ist dies:
> 
> http://<IP-Adresse-der-CCU>/addons/xmlapi/
> 
> 2. Programme aktivieren/deaktivieren
> 
> Programme in Homematic können mit der Script-API über den folgenden Befehl gesteuert werden:
> 
> Aktivieren eines Programms:
> 
> <execute name="ActivateProgram">
>     <command>dom.GetObject('ProgrammName').Active(true);</command>
> </execute>
> 
> Deaktivieren eines Programms:
> 
> <execute name="DeactivateProgram">
>     <command>dom.GetObject('ProgrammName').Active(false);</command>
> </execute>
> 
> 3. Einbettung in HTTP-Requests
> 
> Du kannst ein CURL-Command oder ein HTTP-Tool verwenden, um diese Befehle auszuführen.
> 
> Beispiel mit CURL:
> 
> curl -i -H "Content-Type: text/plain" -d "dom.GetObject('ProgrammName').Active(true);" http://<IP-Adresse-der-CCU>/rega.exe
> 
> Ersetze ProgrammName durch den tatsächlichen Namen deines Programms.
> 
> 4. Nutzung in ioBroker oder anderen Automatisierungen
> 
> Wenn du ioBroker verwendest, kannst du diese Befehle in Skripten oder in Verbindung mit dem Homematic-Adapter einsetzen, um Programme zu steuern.
> 
> 5. Rückmeldung über den Status eines Programms
> 
> Falls du den aktuellen Status eines Programms überprüfen möchtest, kannst du folgendes Script senden:
> 
> <execute name="GetProgramStatus">
>     <command>WriteLine(dom.GetObject('ProgrammName').Active());</command>
> </execute>
> 
> Das gibt den Status des Programms (true oder false) zurück.
> 
> Hinweis
> 
> Stelle sicher, dass dein Homematic-System API-Zugriffe erlaubt und dass du entsprechende Berechtigungen für die Steuerung hast. Die ReGaHSS-API ist standardmäßig aktiviert, aber bei unsicherer Konfiguration kann eine Authentifizierung erforderlich sein.

in ioBroker macht das der sogenannte hm-rega Adapter: https://github.com/ioBroker/ioBroker.hm-rega .

Vielleicht hilft dir das weiter?

**Reply by SukramJ** [reply]
Bitte verschone mich mit der ChatGPT!!!
Entweder Du kannst ein Lösung liefen oder nicht. 
Ich werde Repos nicht nach möglichen Code-Schnippseln durchsuchen.

**Reply by pdmetze** [reply]
Mehr kann ich dir leider nicht geben, wie gesagt bin ich kein Entwickler und habe davon wenig Ahnung. Wenn dir die Beispiele oben nicht helfen, dann ist das leider so.

Nichts für Ungut, aber ich werde mich dann mit Ideen in Zukunft zurückhalten. Habe nicht das Gefühl, dass die bei dir sehr willkommen sind. Trotzdem danke für deine Mühen.

**Reply by SukramJ** [reply]
> Mehr kann ich dir leider nicht geben, wie gesagt bin ich kein Entwickler und habe davon wenig Ahnung. Wenn dir die Beispiele oben nicht helfen, dann ist das leider so.

Vielleich kann jemand mit Ahnung von Skripten was beitragen.

> Nichts für Ungut, aber ich werde mich dann mit Ideen in Zukunft zurückhalten. Habe nicht das Gefühl, dass die bei dir sehr willkommen sind. Trotzdem danke für deine Mühen.

Ideen kannst Du gerne anbringen, aber ChatGPT ist ein NO-GO. 

> Habe nicht das Gefühl, dass die bei dir sehr willkommen sind.

Ich hab einfach keine Lust, das ich bei einer Idee, wo ich ja die Umsetzung machen würde, auch noch die Lösung aus dem Internet rauspopeln soll.

**Reply by SukramJ** [reply]
> .... Habe nicht das Gefühl, dass die bei dir sehr willkommen sind. ...

Damit nicht der falsche Eindruck entsteht: Ich hab recherchiert ob es eine Schnittstelle gibt, oder wie man es umsetzten kann bevor ich dir geantwortet habe, d.h. ich hab mir deine Idee angeschaut und meine Freizeit dafür investiert!!!!  

Also hab auch Verständnis dafür, das mich solche Kommentare echt nerven.

**Comment by SukramJ** [comment]
Done by #1950

**Reply by pdmetze** [reply]
Das ging ja fix, vielen Dank! Ich würde es gerne testen, ich habe jetzt nur nicht gefunden, wie es geht. 

- Ich habe gerade Homematic(IP) Local auf Version 1.76.1 aktualisiert. 
- In der Homematic habe ich "HAHM" in die Description eines Programms hinzugefügt. 
- Ich habe Home Assistant neu gestartet und auch Homematic(IP) Local reloaded. 
- Ich habe die Entität für das Program enabled (und auch nochmal disabled, alles neu gestartet und enabled)

Ich bin sicher einfach zu doof, aber ich finde nicht, wie ich das Programm nun (de)aktivieren kann. Habe ich etwas vergessen? Sorry wie gesagt ich bin recht neu in Home Assistant und lerne immer noch dazu...

![image](https://github.com/user-attachments/assets/5c06f772-78c2-4fbe-a6d7-76659478ed6c)

![image](https://github.com/user-attachments/assets/602d75e6-260d-4293-bb27-82f15692cbb0)

**Reply by SukramJ** [reply]
Du musst bis zum Release warten.

**Reply by dk1305hb** [reply]
@pdmetze, auch wenn wir jetzt hier das Feature bekommen, es ist aus meiner Sicht ratsam sich zu fragen, ob man das nicht schon mit HA selber umsetzen kann. Da gibt es das alles bereits. Wenn du länger mit HA arbeitest kommst du vielleicht dazu deine CCU Programme mit Automations nachzubauen und sich den verkrusteten CCU Strukturen zu entledigen.

**Reply by pdmetze** [reply]
> @pdmetze, auch wenn wir jetzt hier das Feature bekommen, es ist aus meiner Sicht ratsam sich zu fragen, ob man das nicht schon mit HA selber umsetzen kann. Da gibt es das alles bereits. Wenn du länger mit HA arbeitest kommst du vielleicht dazu deine CCU Programme mit Automations nachzubauen und sich den verkrusteten CCU Strukturen zu entledigen.

Hi @dk1305hb, zunächst mal denke ich, ist mit diesem Feature, das SukramJ jetzt implementiert hat, die Integration von Homematic Programmen sinnvoller abgebildet als zuvor. Das sind ja wie Automationen in HA eher triggergesteuerte Logiken und nicht nur manuell angestoßene Befehlsfolgen.

Ob man grundsätzlich alles nach Home Assistant umziehen sollte ist dann ja noch eine andere Frage. Aus meiner Sicht hat das Vor- und Nachteile. Logiken, die ausschließlich Homematic betreffen und sehr essenziell sind, wie z.B. die interne Lichtsteuerung lasse ich lieber in der Homematic, um weniger Abhängigkeiten zu haben; zumal sie auch perfekt funktionieren. Das kann natürlich jeder machen wie er will, für mich passt das in dem Fall so besser. Mit dem Feature jetzt haben wir jedenfalls alle Möglichkeiten und das ist ja schonmal nur ein Vorteil ;-)

**Reply by dk1305hb** [reply]
Das kann ich nachvollziehen. Es ist viel Arbeit mit den Programmen umzuziehen, gerade wenn komplexe fehlerfrei laufen. Wenn ich Programme aktiv/deaktiv steuern wollte hatte ich das in der Bedingung mit einer boolschen Systemvariable gelöst. Deaktiv sind bei mir nur wirklich unbenutzte Programme. Aber das macht jeder wie er es am besten für sich ausmacht :)
Ich hatte beim Lesen einiger Beträge hier im Kopf, dass man nicht mehr zu sehr auf die Software der CCU setzen sollte und neue Sachen lieber im HA direkt umsetzt oder migriert, um zukunftsfähiger zu sein.

**Reply by pdmetze** [reply]
Ja. Hardware Performance ist da ja so ein Thema bei der CCU. Ich habe damals mit der CCU1 angefangen, da konnte man echt nicht viel machen. Die CCU3 ist zwar deutlich besser, aber wirklich viel sollte man darin auch nicht laufen lassen. Bei mir sind nur essenzielle und interne Sachen in der Homematic. Ich werde mir deinen Ratschlag auf jeden Fall beherzigen, alles weitere und neue kommt in Home Assistant.

Deine Lösung mit extra System-Variablen wäre mein Plan B gewesen ;-) So ist es jetzt aber viel einfacher und man spart sich die extra Systemvariablen und Bedingungen in den Programmen.

**Reply by dk1305hb** [reply]
Das ist schön zu lesen, komme auch von der CCU1. Da haben wir schon einiges durch :)
